pipenv/cli/options.py

```diff
@@ -0,0 +1,385 @@
+        state = ctx.ensure_object(State)
+        state.installstate.keep_outdated = value
+        return value
+    return option("--keep-outdated", is_flag=True, default=False, expose_value=False,
+                    help=u"Keep out-dated dependencies from being updated in Pipfile.lock.",
+                    callback=callback, type=click_booltype)(f)
+
+
+def selective_upgrade_option(f):
+    def callback(ctx, param, value):
+        state = ctx.ensure_object(State)
+        state.installstate.selective_upgrade = value
+        return value
+    return option("--selective-upgrade", is_flag=True, default=False, type=click_booltype,
+                    help="Update specified packages.", callback=callback,
+                    expose_value=False)(f)
+
+
+def ignore_pipfile_option(f):
+    def callback(ctx, param, value):
+        state = ctx.ensure_object(State)
+        state.installstate.ignore_pipfile = value
+        return value
+    return option("--ignore-pipfile", is_flag=True, default=False, expose_value=False,
+                    help="Ignore Pipfile when installing, using the Pipfile.lock.",
+                    callback=callback)(f)
+
+
+def dev_option(f):
+    def callback(ctx, param, value):
+        state = ctx.ensure_object(State)
+        state.installstate.dev = value
+        return value
+    return option("--dev", "-d", is_flag=True, default=False, type=click_booltype,
+                    help="Install package(s) in [dev-packages].", callback=callback,
+                    expose_value=False)(f)
+
+
+def pre_option(f):
+    def callback(ctx, param, value):
+        state = ctx.ensure_object(State)
+        state.installstate.pre = value
+        return value
+    return option("--pre", is_flag=True, default=False, help=u"Allow pre-releases.",
+         callback=callback, type=click_booltype, expose_value=False)(f)
+
+
+def package_arg(f):
+    def callback(ctx, param, value):
+        state = ctx.ensure_object(State)
+        state.installstate.packages.extend(value)
+        return value
+    return argument('packages', nargs=-1, callback=callback, expose_value=False,)(f)
+
+
+def three_option(f):
+    def callback(ctx, param, value):
+        state = ctx.ensure_object(State)
+        if value is not None:
+            state.three = value
+            state.two = not value
+        return value
+    return option("--three/--two", is_flag=True, default=None,
+                    help="Use Python 3/2 when creating virtualenv.", callback=callback,
+                    expose_value=False)(f)
+
+
+def python_option(f):
+    def callback(ctx, param, value):
+        state = ctx.ensure_object(State)
+        if value is not None:
+            state.python = validate_python_path(ctx, param, value)
+        return value
+    return option("--python", default=False, nargs=1, callback=callback,
+                    help="Specify which version of Python virtualenv should use.",
+                    expose_value=False)(f)
```

Comment:
I hope we can merge `two`, `three` and `python` soon, but it probably needs to be done after this is merged.

Reply:
I did consider that, but actually they are grouped together in `common_options` and don't appear anywhere else anyway.  I tried my best to keep this organized and grouped based on how we are using these things currently.  The real concern here (and the thing to watch out for in this review) is that the state passing approach I *think* will pass the cli options and args as well... so we just need to be sure this doesn't accidentally add new CLI options to the API